include/deal.II/base/mpi_noncontiguous_partitioner.templates.h

```diff
@@ -321,17 +323,25 @@ namespace Utilities
           AssertThrowMPI(ierr);
         }
 
-      auto src_iterator = src.begin();
-
       // post send
       for (types::global_dof_index i = 0, k = 0; i < send_ranks.size(); i++)
         {
           // collect data to be send
+          AssertIndexRange(i + 1, send_ptr.size());
           for (types::global_dof_index j = send_ptr[i]; j < send_ptr[i + 1];
                j++)
-            buffers[j] = src_iterator[send_indices[k++]];
+            {
+              AssertIndexRange(k, send_indices.size());
```

Comment:
```suggestion
              AssertIndexRange(send[ptr[i + 1], send_indices.size() + 1);
```

Reply:
This may be true but I don't see why - in particular, I don't see why `send_ptr[0] = 0`, which is necessary for this substitution to work. I'd rather keep the less sophisticated check here so that we don't rely on nonlocal assumptions.

Reply:
Yes, I agree. I was thinking you were checking for `j` and not for `k`.